processor.go

```diff
@@ -143,7 +130,214 @@ func (p *ServiceProcessor) RegisterStreamingHandler(f interface{}) error {
+	}
+	if in0.NumField() == 0 {
+		return emptyGET, "", nil
+	} else if in0.NumField() == 1 {
+		// we support int and byte slices only
+		if in0.Field(0).Type.Kind() == reflect.Slice && in0.Field(0).Type.Elem().Kind() == reflect.Uint8 {
+			return sliceGET, in0.Field(0).Name, nil
+		} else if in0.Field(0).Type.Kind() == reflect.Int {
+			return intGET, in0.Field(0).Name, nil
+		}
+		return invalidGET, "", errors.New("only byte slices and int are supported")
+	}
+	return invalidGET, "", errors.New("number of fields must be 0 or 1")
+}
+
+// RegisterRESTHandler takes a callback of type
+// func(msg interface{})(ret interface{}, err error),
+// where msg and ret must be pointers to structs.
+// For POST and PUT, the callback is registered on the URL
+// /v$version/$namespace/$msgStructName. The client should serialize the request
+// using JSON and set the conent type to application/json to use the service.
+// The response is also JSON encoded.
+//
+// For GET requests, the callback is registered on the same URL. But clients
+// can also query individual resources such as
+// /v$version/$namespace/$msgStructName/$id. For this to work, msg in the
+// callback must be a singleton struct with either an integer or a byte slice.
+// For integers, the client can directly query the integer resource, for byte
+// slices, the clients must query the hex encoded representation. Using an
+// empty struct for msg is also supported.
+//
+// The min/maxVersion argument represents the range of versions where the API
+// is present. If breaking changes must be made then they must use a new
+// version.
+//
+// This method is experimental.
+func (p *ServiceProcessor) RegisterRESTHandler(f interface{}, namespace, method string, minVersion, maxVersion int) error {
+	// TODO support more methods
+	if method != "GET" && method != "POST" && method != "PUT" {
+		return errors.New("invalid REST method")
+	}
+	if minVersion > maxVersion {
+		return errors.New("min version is greater than max version")
+	}
+	if minVersion < 3 {
+		return errors.New("earliest supported API level must be greater or equal to 3")
+	}
+	if err := handlerInputCheck(f); err != nil {
+		return err
+	}
+	resource, sh, err := createServiceHandler(f)
+	if err != nil {
+		return err
+	}
+	var k kindGET
+	var fieldNameGET string
+	if method == "GET" {
+		k, fieldNameGET, err = prepareHandlerGET(f)
+		if err != nil {
+			return err
+		}
+	}
+	h := func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != method {
+			http.Error(w, "unsupported method: "+r.Method, http.StatusMethodNotAllowed)
+			return
+		}
+		var msgBuf []byte
+		switch r.Method {
+		case "GET":
+			if k == emptyGET {
+				msgBuf = []byte("{}")
+			} else if k == intGET {
+				ok, err := regexp.MatchString(fmt.Sprintf(`^/v\d/%s/%s/\d+$`, namespace, resource), r.URL.EscapedPath())
+				if err != nil {
+					http.Error(w, err.Error(), http.StatusInternalServerError)
```

Comment:
5 levels of indentation is quite a lot, would it be worth moving some part to a function?

Reply:
the high level of indentation is indeed not i deal, so I tried to refactor a bit. but I think moving the code around, either the code inside `case "GET"`, or the handler code, to another function is a bit messy because i have to pass a few arguments around. OTOH, to make the code a bit nicer, i've made the code width shorter in the latest commit.